server/jetstream_cluster.go

```diff
@@ -1621,9 +1677,61 @@ func (js *jetStream) monitorStream(mset *stream, sa *streamAssignment, sendSnaps
 				js.setStreamAssignmentRecovering(sa)
 			}
 
+			// Process our leader change.
 			js.processStreamLeaderChange(mset, isLeader)
+
+			// Check for migrations here. We set the state on the stream assignment update below.
+			if isLeader && migrating {
+				if peerGroup == oldPeerGroup {
+					startMigrationMonitoring()
+				} else {
+					stopMigrationMonitoring()
+					// We delay here since consumer migration, while in most cases is faster, is done after the stream itself.
+					// When we remove the stream from the old peers this will remove the consumers which might still be in flight.
+					time.AfterFunc(1000*time.Millisecond, func() { js.removeOldPeers(mset) })
```

Comment:
Not sure I understand the comment, but that sounds a bit scary: do we have to be sure that consumer migration is done when removeOldPeers is invoked (in 1sec)? You can't assume how long things are going to take.

Other concern: since this is happening based on an anonymous timer (that can't be stopped), is there a risk that this code here executed (scheduling the peers removal) followed within 1sec by a start migration? When the timer's function finally executes, is that not going to mess up the state of the newly started migration?

Reply:
I meant to lower it to 500ms, which I will do. 

This is when we have moved control to the new stream leader in the new peerset. It is responsible to shrink the peerset, but consumers have to be expanded behind the stream itself since the consumer needs the stream state on any peer it occupies.

In some of my testing I saw this drastically slow down the consumer checks since the stream removed the old consumers and in some instances the old leader before it could transfer leadership to the new peerset etc.

Reply:
I lowered, hopefully the explanation makes sense.